File: com/white/MaxAreaOfIsland.java
package com.white;

public class MaxAreaOfIsland {

    // problem 695

    private static int[][] inside;

    public static int maxAreaOfIsland(int[][] grid) {
        inside = grid;
        int max = 0;
        int temp;
        for (int i=0; i<grid[0].length; i++) {
            for (int j=0; j<grid.length; j++) {
                if (grid[j][i] == 1) {
                    temp = wanderer(i,j);
                    max = (max>temp) ? max : temp;
                }
            }
        }
        return max;
    }

    public static int wanderer(int i, int j) {
        if (inside[j][i]==0) {
            return 0;
        } else {
            inside[j][i] = 0;
            return 1 + wanderer(Math.min(inside[0].length-1,i+1),j) + wanderer(Math.max(0,i-1),j) + wanderer(i,Math.min(inside.length-1,j+1)) + wanderer(i,Math.max(0,j-1));
        }
    }


}

File: LeetcodeCore/MaxAreaOfIsland.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace LeetcodeCore
{
    public class MaxAreaOfIslandSolution
    {
        // 695. Max Area of Island
        public int MaxAreaOfIsland(int[][] grid)
        {
            var m = grid.Length;
            var n = grid[0].Length;
            var totalLength = m * n;
            var uf = new UnionFind(totalLength + 1);
            var maxArea = 0;

            for (int i = 0; i < m; i++)
            {
                for (int j = 0; j < n; j++)
                {
                    if (grid[i][j] == 0)
                    {
                        uf.Union(i * n + j, totalLength);
                    }
                    else
                    {
                        if (j + 1 < n && grid[i][j + 1] == 1)
                        {
                            uf.Union(i * n + j, i * n + j + 1);
                        }
                        if (i + 1 < m && grid[i + 1][j] == 1)
                        {
                            uf.Union((i + 1) * n + j, i * n + j);
                        }
                        maxArea = Math.Max(maxArea, uf.FindSize(i * n + j));
                    }
                }
            }

            return maxArea;
        }
    }
}
